### PR TITLE
infoschema/slow_query: fix token too long

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
This PR try to fix If there is row length more than `bufio.MaxScanTokenSize`, parse slow log will get error: `bufio.Scanner: token too long`.

### What is changed and how it works?
Adjust `bufio.Buffer` before parse slow log and refine error msg.

* Add `QueryLogMaxLenRecord` to record the max `QueryLogMaxLen`.

**Attention**:  If the user `set @@tidb_query_log_max_len=65536` and write a long slow query to slow log, then the tidb-restart. Because the `QueryLogMaxLenRecord` doesn't persistent Stores, so it will be the 
`QueryLogMaxLen` default value `2048`, and then query `INFORMATION_SCHEMA.SLOW_QUERY` will also get error: `"read file buffer overflow, please try to enlarge the variable 'tidb_query_log_max_len'`, but I think the error message is more clear now, and just enlarge `tidb_query_log_max_len` then retry.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by crazycs520** [comment]
@eurekaka @winkyao PTAL

File: infoschema/slow_log.go [hunk 1]
```diff
@@ -88,6 +90,11 @@ func ParseSlowLog(tz *time.Location, scanner *bufio.Scanner) ([][]types.Datum, e
 	startFlag := false
 	var st *slowQueryTuple
 	var err error
+	// Adjust buffer size.
+	if atomic.LoadUint64(&config.QueryLogMaxLenRecord) > (bufio.MaxScanTokenSize - 100) {
+		maxBuf := int(atomic.LoadUint64(&config.QueryLogMaxLenRecord) + 100)
+		scanner.Buffer([]byte{}, maxBuf)
+	}
```
Comment (eurekaka) [inline infoschema/slow_log.go]: Can we use `bufio.Reader.ReadLine` to completely avoid this error?

Comment (winkyao) [reply infoschema/slow_log.go]: We can use ReadLine to implement a new scanner

Comment (crazycs520) [reply infoschema/slow_log.go]: Great idea. Other question, should we add a max-single-line-length-limit? If user executes a long query. 

At first, I think if we don't limit the max-single-line-length, then parse slow log may consume a lot of memory and  then may make TiDB-server killed by the system.

But after careful consideration, I think no need to add max-single-line-length, the first reason is we have log rotate, the other reason is executing the long query may make tidb-server killed in executing. 😂  What do you think? @eurekaka @winkyao

Comment (eurekaka) [reply infoschema/slow_log.go]: https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_max_allowed_packet should limit the input query size?

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 
https://github.com/pingcap/tidb/issues/10340

File: infoschema/slow_log.go (hunk 1 above)
Comment (crazycs520) [reply infoschema/slow_log.go]: done. PTAL @eurekaka @winkyao

**Comment by crazycs520** [comment]
@eurekaka PTAL

File: infoschema/slow_log.go [hunk 2]
```diff
@@ -127,12 +138,30 @@ func ParseSlowLog(tz *time.Location, scanner *bufio.Scanner) ([][]types.Datum, e
...
+	var tempLine []byte
+	if err != nil {
+		return lineByte, err
+	}
+	for isPrefix {
+		tempLine, isPrefix, err = reader.ReadLine()
+		lineByte = append(lineByte, tempLine...)
+
+		// Use the max value of max_allowed_packet to check the single line length.
+		if len(lineByte) > int(variable.MaxOfMaxAllowedPacket) {
```
Comment (eurekaka) [inline infoschema/slow_log.go]: Why using MaxOfMaxAllowedPacket instead of MaxAllowedPacket?

File: infoschema/slow_log.go [hunk 3]
```diff
@@ -77,19 +78,29 @@ func parseSlowLogFile(tz *time.Location, filePath string) ([][]types.Datum, erro
...
-	for scanner.Scan() {
-		line := scanner.Text()
+	finish := false
+	for {
+		if finish {
+			break
+		}
+		lineByte, err := getOneLine(reader)
+		if err == io.EOF {
+			finish = true
```
Comment (tiancaiamao) [inline infoschema/slow_log.go]: If err == io.EOF, should we break this loop immediately ?

File: infoschema/slow_log.go (hunk 2 above)
Comment (crazycs520) [reply infoschema/slow_log.go]: Because `MaxAllowedPacket` may has been changed.

File: infoschema/slow_log.go (hunk 3 above)
Comment (crazycs520) [reply infoschema/slow_log.go]: done.

**Pull request merged**